### PR TITLE
Add context gettable and settable calls

### PR DESCRIPTION
**Pull request by paulidale**
This permits context specific param arrays to be returned.

A fallback mechanism is also implemented to call the non-context specific version is the specific ones are not available.

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/evp/digest.c [hunk 1]
```diff
@@ -678,9 +678,14 @@ const OSSL_PARAM *EVP_MD_CTX_settable_params(EVP_MD_CTX *ctx)
...
                    pctx->op.sig.sigprovctx);
 
-    if (ctx->digest != NULL && ctx->digest->settable_ctx_params != NULL)
-        return ctx->digest->settable_ctx_params(
-                  ossl_provider_ctx(EVP_MD_provider(ctx->digest)));
+    if (ctx->digest != NULL) {
+        void *prov_ctx = ossl_provider_ctx(EVP_MD_provider(ctx->digest));
+
+        if (ctx->digest->ctx_settable_params != NULL)
+            return ctx->digest->ctx_settable_params(prov_ctx, ctx->provctx);
```
Comment (levitte) [inline crypto/evp/digest.c:685]: In hindsight, `ctx->provctx` isn't the best name...

File: doc/man7/provider-cipher.pod [hunk 2]
```diff
@@ -42,6 +42,10 @@ provider-cipher - The cipher library E<lt>-E<gt> provider functions
  /* Cipher operation parameter descriptors */
  const OSSL_PARAM *OSSL_FUNC_cipher_gettable_ctx_params(void *provctx);
  const OSSL_PARAM *OSSL_FUNC_cipher_settable_ctx_params(void *provctx);
+ const OSSL_PARAM *OSSL_FUNC_cipher_ctx_gettable_params(void *provctx,
+                                                        void *cctx);
+ const OSSL_PARAM *OSSL_FUNC_cipher_ctx_settable_params(void *provctx,
+                                                        void *cctx);
```
Comment (levitte) [inline doc/man7/provider-cipher.pod:48]: From a provider author perspective, I find it quite clumsy to have two functions that essentially do the same thing. I'd rather see them merged into one, and being able to vary the response depending on if the second argument is NULL or not.

File: crypto/evp/digest.c (hunk 1 above)
Comment (paulidale) [reply crypto/evp/digest.c:685]: Agreed.

File: doc/man7/provider-cipher.pod (hunk 2 above)
Comment (paulidale) [reply doc/man7/provider-cipher.pod:48]: A provider author need only implement the existing function and everything will be happy.  The new function is only required if the parameters can change based on the state of the context.

Not sure if this justifies separate functions or not but it isn't onerous or clumsy for provider authors.

File: crypto/evp/mac_meth.c [hunk 3]
```diff
@@ -193,6 +205,28 @@ const OSSL_PARAM *EVP_MAC_settable_ctx_params(const EVP_MAC *mac)
     return mac->settable_ctx_params(ossl_provider_ctx(EVP_MAC_provider(mac)));
 }
 
+const OSSL_PARAM *EVP_MAC_CTX_gettable_params(EVP_MAC_CTX *ctx)
+{
+    void *prov_ctx = ossl_provider_ctx(EVP_MAC_provider(ctx->meth));
+
+    if (ctx->meth->ctx_gettable_params != NULL)
+        return ctx->meth->ctx_gettable_params(prov_ctx, ctx->data);
```
Comment (slontis) [inline crypto/evp/mac_meth.c:213]: why do this if we dont resolve it?

File: doc/man3/EVP_EncryptInit.pod [hunk 4]
```diff
@@ -881,8 +890,9 @@ disappeared.  EVP_CIPHER_CTX_init() remains as an alias for
 EVP_CIPHER_CTX_reset().
 
 The EVP_CIPHER_fetch(), EVP_CIPHER_free(), EVP_CIPHER_up_ref(),
-EVP_CIPHER_CTX_set_params() and EVP_CIPHER_CTX_get_params() functions
-were added in 3.0.
+EVP_CIPHER_CTX_set_params(), EVP_CIPHER_CTX_get_params(),
+EVP_CIPHER_CTX_gettable_params() and EVP_CIPHER_CTX_settable_params()
```
Comment (slontis) [inline doc/man3/EVP_EncryptInit.pod:894]: EVP_CIPHER_settable_ctx_params?

File: doc/man3/EVP_KDF.pod [hunk 5]
```diff
@@ -124,11 +127,16 @@ simply ignored.
...
-and EVP_KDF_CTX_set_params(), respectively.
+EVP_KDF_gettable_params() returns parameters that can be used with
+EVP_KDF_get_params().  See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM>
+as parameter descriptor.
+
+EVP_KDF_gettable_ctx_params() and EVP_KDF_CTX_gettable_params()
+return parameters that can be used with EVP_KDF_CTX_get_params().
+See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as parameter descriptor.
+
+EVP_KDF_settable_ctx_params() and EVP_KDF_CTX_settable_params()
```
Comment (slontis) [inline doc/man3/EVP_KDF.pod:138]: This doesnt tell me what the diff between settable_ctx_params() and CTX_settable_params() is?

File: doc/man7/provider-cipher.pod (hunk 2 above)
Comment (slontis) [reply doc/man7/provider-cipher.pod:48]: I think the one API would be less confusing..

**Review by slontis — COMMENTED** [review]
Remember to add to the openssl app also (I know its WIP).

**Comment by paulidale** [comment]
Add what to the app????

File: crypto/evp/mac_meth.c (hunk 3 above)
Comment (paulidale) [reply crypto/evp/mac_meth.c:213]: ??? I don't understand

Comment (slontis) [reply crypto/evp/mac_meth.c:213]: The resolve does this
````
        case OSSL_FUNC_MAC_CTX_GETTABLE_PARAMS:
            if (mac->ctx_gettable_params != NULL)
                break;
````

So why check later if it is NULL and call it when it is always going to be NULL.

Comment (paulidale) [reply crypto/evp/mac_meth.c:213]: The resolve is checking for a double definition.  If it isn't defined at all it can still be NULL, so there needs to be a check here.

**Comment by slontis** [comment]
> Add what to the app????



> Add what to the app????
e.g.
````
./apps/openssl list -cipher-algorithms -verbose
````

File: doc/man7/provider-cipher.pod (hunk 2 above)
Comment (paulidale) [reply doc/man7/provider-cipher.pod:48]: Major reworking coming up.  Most of what's been done here is not relevant with this change.

**Comment by paulidale** [comment]
I'm not proposing to create a context for every algorithm in order to query settable/gettable parameters.  The algorithm call should be sufficient.

File: doc/man7/provider-cipher.pod (hunk 2 above)
Comment (paulidale) [reply doc/man7/provider-cipher.pod]: I've switched to a single libcrypto ↔ provider API that handles both flavours.  Comments welcomed.

**Review by slontis — COMMENTED** [review]
Maybe adding code for a case that actually uses this mechanism.
Otherwise it looks ok to me..

**Comment by paulidale** [comment]
The only cases we've currently got are dealt with at the EVP layer.
I can see KDFs and MACs using this in due course.

File: crypto/evp/digest.c [hunk 6]
```diff
@@ -680,7 +681,7 @@ const OSSL_PARAM *EVP_MD_CTX_settable_params(EVP_MD_CTX *ctx)
 
     if (ctx->digest != NULL && ctx->digest->settable_ctx_params != NULL)
         return ctx->digest->settable_ctx_params(
-                  ossl_provider_ctx(EVP_MD_provider(ctx->digest)));
+                ctx->provctx, ossl_provider_ctx(EVP_MD_provider(ctx->digest)));
```
Comment (slontis) [inline crypto/evp/digest.c:684]: This indent is pretty horrible.

File: doc/man7/provider-digest.pod [hunk 7]
```diff
@@ -152,11 +154,16 @@ Any parameter settings are additional to any that were previously set.
 OSSL_FUNC_digest_get_ctx_params() gets digest operation details details from
 the given provider side digest context I<dctx> and stores them in I<params>.
 
-OSSL_FUNC_digest_gettable_params(), OSSL_FUNC_digest_gettable_ctx_params(), and
-OSSL_FUNC_digest_settable_ctx_params() all return constant B<OSSL_PARAM> arrays
-as descriptors of the parameters that OSSL_FUNC_digest_get_params(),
-OSSL_FUNC_digest_get_ctx_params(), and OSSL_FUNC_digest_set_ctx_params() can handle,
-respectively.
+OSSL_FUNC_digest_gettable_params() returns a constant B<OSSL_PARAM> array
+containing desriptors of the parameters that OSSL_FUNC_digest_get_params()
```
Comment (slontis) [inline doc/man7/provider-digest.pod:158]: descriptors

File: doc/man7/provider-digest.pod [hunk 8]
```diff
@@ -152,11 +154,16 @@ Any parameter settings are additional to any that were previously set.
...
+can handle.
+
+OSSL_FUNC_digest_gettable_ctx_params() and
+OSSL_FUNC_digest_settable_ctx_params() both return constant
+B<OSSL_PARAM> arrays as descriptors of the parameters that
+OSSL_FUNC_digest_get_ctx_params() and OSSL_FUNC_digest_set_ctx_params()
+can handle, respectively.  The array is based on the current state of the
+context if I<dctx> is not NULL and on the algorithm I<provctx> otherwise.
```
Comment (slontis) [inline doc/man7/provider-digest.pod:166]: digest side provider context
algorithm ??

File: doc/man3/EVP_DigestInit.pod [hunk 9]
```diff
@@ -180,18 +180,29 @@ See L</PARAMETERS> below for more information.
...
-EVP_MD_CTX_gettable_params() return parameters that can be used with
-EVP_MD_CTX_get_params(). EVP_MD_settable_ctx_params() and
-EVP_MD_CTX_settable_params() return parameters that can be used with
-EVP_MD_CTX_set_params().
-See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as parameter descriptor.
+=item EVP_MD_gettable_params()
+
+Get a constant B<OSSL_PARAM> array that describes the retrievable parameters
+that can be used with EVP_MD_get_params().  See L<OSSL_PARAM(3)> for the
+use of B<OSSL_PARAM> as parameter descriptor.
```
Comment (slontis) [inline doc/man3/EVP_DigestInit.pod:187]: as a?

File: doc/man3/EVP_DigestInit.pod [hunk 10]
```diff
@@ -180,18 +180,29 @@ See L</PARAMETERS> below for more information.
...
+
+=item EVP_MD_gettable_ctx_params(), EVP_MD_CTX_gettable_params()
+
+Get a constant B<OSSL_PARAM> array that describes the retrievable parameters
+that can be used with EVP_MD_CTX_get_params().  EVP_MD_gettable_ctx_params()
+returns the parameters that can be retrieved from the algorithm, whereas
+EVP_MD_CTX_gettable_params() returns the parameters that can be retrieved
+in the context's current state.  See L<OSSL_PARAM(3)> for the use of
+B<OSSL_PARAM> as parameter descriptor.
```
Comment (slontis) [inline doc/man3/EVP_DigestInit.pod:196]: as a?

File: doc/man3/EVP_DigestInit.pod [hunk 11]
```diff
@@ -180,18 +180,29 @@ See L</PARAMETERS> below for more information.
...
+
+=item EVP_MD_settable_ctx_params(), EVP_MD_CTX_settable_params()
+
+Get a constant B<OSSL_PARAM> array that describes the settable parameters
+that can be used with EVP_MD_CTX_set_params().  EVP_MD_settable_ctx_params()
+returns the parameters that can be set from the algorithm, whereas
+EVP_MD_CTX_settable_params() returns the parameters that can be set in the
+context's current state.  See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM>
+as parameter descriptor.
```
Comment (slontis) [inline doc/man3/EVP_DigestInit.pod:205]: ..

File: doc/man3/EVP_DigestInit.pod (hunk 10 above)
Comment (slontis) [reply doc/man3/EVP_DigestInit.pod:196]: Note this is duplicated in other files.

File: crypto/evp/digest.c (hunk 6 above)
Comment (paulidale) [reply crypto/evp/digest.c:684]: What do you suggest?  It's too long...

File: doc/man7/provider-digest.pod (hunk 7 above)
Comment (paulidale) [reply doc/man7/provider-digest.pod:158]: Fixed.

File: doc/man3/EVP_DigestInit.pod (hunk 9 above)
Comment (paulidale) [reply doc/man3/EVP_DigestInit.pod:187]: Damn pre-existing wording problems.....

File: crypto/evp/digest.c (hunk 6 above)
Comment (slontis) [reply crypto/evp/digest.c:684]: Maybe

````
{
   blarg *b = a;
   return fgggjkkgdkgd(b);
}
````

File: doc/man3/EVP_DigestInit.pod (hunk 9 above)
Comment (paulidale) [reply doc/man3/EVP_DigestInit.pod:187]: Fixed.

File: doc/man7/provider-digest.pod (hunk 8 above)
Comment (paulidale) [reply doc/man7/provider-digest.pod:166]: Poor choice of naming.  `provctx` refers to the algorithm returned by the provider to libcrypto.  @levitte [noted this above](https://github.com/openssl/openssl/pull/14240#discussion_r578879244).

I'm not proposing to change this one globally.

File: crypto/evp/digest.c (hunk 6 above)
Comment (levitte) [reply crypto/evp/digest.c:684]: @slontis made a good suggestion, please do that

Comment (slontis) [reply crypto/evp/digest.c:684]: There are a few more of these indents in this PR.

Comment (levitte) [reply crypto/evp/digest.c:684]: Yes, I hope it's understood to be a broad recommendation and not just for this spot.

File: doc/man7/provider-digest.pod (hunk 8 above)
Comment (levitte) [reply doc/man7/provider-digest.pod]: Agreed, let's make that in another PR.  The MAC API uses `data` instead, which turns out to be much less confusing.

Comment (paulidale) [reply doc/man7/provider-digest.pod]: +1 for `data`

File: crypto/evp/digest.c (hunk 6 above)
Comment (paulidale) [reply crypto/evp/digest.c:684]: They are all pre-existing indentation 😞

Comment (levitte) [reply crypto/evp/digest.c:684]: That sucks, of course...  should have been caught earlier.

File: doc/man7/provider-digest.pod (hunk 8 above)
Comment (levitte) [reply doc/man7/provider-digest.pod]: Someone raise the issue?

File: crypto/evp/digest.c (hunk 6 above)
Comment (slontis) [reply crypto/evp/digest.c:684]: More than likely I probably did it :)

Comment (paulidale) [reply crypto/evp/digest.c]: Fixed now.

File: doc/man7/provider-digest.pod (hunk 8 above)
Comment (paulidale) [reply doc/man7/provider-digest.pod]: Issue #14284 raised.

**Comment by paulidale** [comment]
I'm not sure what to do with the outstanding asymmetric calls:
``` C
const OSSL_PARAM *EVP_PKEY_CTX_gettable_params(EVP_PKEY_CTX *ctx);
const OSSL_PARAM *EVP_SIGNATURE_gettable_ctx_params(const EVP_SIGNATURE *sig);
const OSSL_PARAM *EVP_ASYM_CIPHER_gettable_ctx_params(const EVP_ASYM_CIPHER *ciph);
const OSSL_PARAM *EVP_KEYEXCH_gettable_ctx_params(const EVP_KEYEXCH *keyexch);
const OSSL_PARAM *EVP_KEM_gettable_ctx_params(const EVP_KEM *kem);
```

**Comment by paulidale** [comment]
All the symmetric algorithms are done, putting this up for review.

I don't think that the asymmetric algorithms need any changes.  If they do, it'll be a separate PR.

**Comment by paulidale** [comment]
Ping for review?

File: crypto/evp/evp_rand.c [hunk 12]
```diff
@@ -431,15 +431,31 @@ const OSSL_PARAM *EVP_RAND_gettable_ctx_params(const EVP_RAND *rand)
     if (rand->gettable_ctx_params == NULL)
         return NULL;
     return rand->gettable_ctx_params(
-               ossl_provider_ctx(EVP_RAND_provider(rand)));
+               NULL, ossl_provider_ctx(EVP_RAND_provider(rand)));
```
Comment (slontis) [inline crypto/evp/evp_rand.c:434]: you missed the ones in this file :) Dont hate me..

File: doc/man3/EVP_EncryptInit.pod [hunk 13]
```diff
@@ -147,6 +149,8 @@ EVP_CIPHER_do_all_provided
  const OSSL_PARAM *EVP_CIPHER_gettable_params(const EVP_CIPHER *cipher);
  const OSSL_PARAM *EVP_CIPHER_settable_ctx_params(const EVP_CIPHER *cipher);
  const OSSL_PARAM *EVP_CIPHER_gettable_ctx_params(const EVP_CIPHER *cipher);
+ const OSSL_PARAM *EVP_CIPHER_CTX_settable_params(EVP_CIPHER_CTX *ctx);
```
Comment (slontis) [inline doc/man3/EVP_EncryptInit.pod:152]: Add to the history section..

File: doc/man3/EVP_DigestInit.pod [hunk 14]
```diff
@@ -180,18 +180,29 @@ See L</PARAMETERS> below for more information.
...
-EVP_MD_CTX_settable_params() return parameters that can be used with
-EVP_MD_CTX_set_params().
-See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as parameter descriptor.
+=item EVP_MD_gettable_params()
+
+Get a constant B<OSSL_PARAM> array that describes the retrievable parameters
+that can be used with EVP_MD_get_params().  See L<OSSL_PARAM(3)> for the
+use of B<OSSL_PARAM> as a parameter descriptor.
+
+=item EVP_MD_gettable_ctx_params(), EVP_MD_CTX_gettable_params()
```
Comment (slontis) [inline doc/man3/EVP_DigestInit.pod:189]: These should all be in the history section

File: doc/man3/EVP_MAC.pod [hunk 15]
```diff
@@ -47,6 +48,8 @@ EVP_MAC_do_all_provided - EVP MAC routines
  const OSSL_PARAM *EVP_MAC_gettable_params(const EVP_MAC *mac);
  const OSSL_PARAM *EVP_MAC_gettable_ctx_params(const EVP_MAC *mac);
  const OSSL_PARAM *EVP_MAC_settable_ctx_params(const EVP_MAC *mac);
+ const OSSL_PARAM *EVP_MAC_CTX_gettable_params(EVP_MAC_CTX *ctx);
```
Comment (slontis) [inline doc/man3/EVP_MAC.pod:51]: should the ctx be const for these?

File: providers/implementations/ciphers/cipher_aes_siv.c [hunk 16]
```diff
@@ -248,7 +250,6 @@ static OSSL_FUNC_cipher_update_fn lc##_stream_update;                          \
 static OSSL_FUNC_cipher_final_fn lc##_stream_final;                            \
 static OSSL_FUNC_cipher_cipher_fn lc##_cipher;                                 \
 static OSSL_FUNC_cipher_get_params_fn alg##_##kbits##_##lc##_get_params;       \
-static OSSL_FUNC_cipher_gettable_params_fn alg##_##lc##_gettable_ctx_params;   \
```
Comment (slontis) [inline providers/implementations/ciphers/cipher_aes_siv.c:251]: ? why is this removed?

**Review by slontis — COMMENTED** [review]
It looks like all the XX_CTX_settable_params and gettable_params should pass const parameters?

File: crypto/evp/evp_rand.c (hunk 12 above)
Comment (paulidale) [reply crypto/evp/evp_rand.c:434]: :(

File: doc/man3/EVP_EncryptInit.pod (hunk 13 above)
Comment (paulidale) [reply doc/man3/EVP_EncryptInit.pod:152]: And a number of other existing functions :(

File: doc/man3/EVP_DigestInit.pod (hunk 14 above)
Comment (paulidale) [reply doc/man3/EVP_DigestInit.pod:189]: Plus some others....

File: doc/man3/EVP_MAC.pod (hunk 15 above)
Comment (paulidale) [reply doc/man3/EVP_MAC.pod:51]: I don't believe so.  If a context does have a dynamic selection of gettable/settable parameters, it will need to cache the list once queried.  Thus, the structure's contents could be mutable.

File: providers/implementations/ciphers/cipher_aes_siv.c (hunk 16 above)
Comment (paulidale) [reply providers/implementations/ciphers/cipher_aes_siv.c:251]: It's a mistake.  It's a `gettable_params` callback (on the EVP_CIPHER) returning the `gettable_ctx_params` parameters (for the EVP_CIPHER_CTX`).

A proper version of this function wasn't defined or used anywhere.

**Comment by paulidale** [comment]
> It looks like all the XX_CTX_settable_params and gettable_params should pass const parameters?

They could, but they'd need to cast away the const in order to keep a reference to a list they dynamically allocate.  We currently don't allocate such a list (i.e. everything is const), but the entire point of this PR is to permit dynamic parameter lists.  I.e. they need to be mutable.

**Comment by paulidale** [comment]
Comments addressed.

**Comment by t8m** [comment]
> It looks like all the XX_CTX_settable_params and gettable_params should pass const parameters?

IMO this is really unneeded as the callers do not expect the CTX kind of structures to be constant. So even if we could mark it const for some of the cases it probably would not be a good idea as it would prevent doing some caching stuff etc. later without casting away the const.

**Review by t8m — APPROVED** [review]
LGTM, good work!

**Comment by slontis** [comment]
@paulidale What was the last commit here - the force push obscures what happened after approval. A comment helps with this.

**Comment by slontis** [comment]
Your commits with the name 'upport' might be missing a character. If these are going to be seperate commits you might want to rebase/reword.

**Comment by paulidale** [comment]
The post approval push was a rebase that squashed the fixups.  Two rogue fixups that still appear in the commits aren't really fixups.

**Review by slontis — APPROVED** [review]
Approved subject to commit renaming.

**Comment by paulidale** [comment]
And merged!  Thanks for the helpful reviews
